File: telegram_bot/crypto_utils.py
import base64
import os

from Crypto.Cipher import AES
from Crypto.Util.Padding import pad, unpad
from dotenv import load_dotenv

load_dotenv()

# Load and validate key
key_b64 = os.getenv("AES_SECRET_KEY")
if not key_b64:
    raise ValueError("Missing AES_SECRET_KEY in .env")

KEY = base64.b64decode(key_b64)
if len(KEY) != 32:
    raise ValueError("AES key must be 32 bytes")

# Load and validate fixed IV
iv_b64 = os.getenv("FIXED_IV")
if not iv_b64:
    raise ValueError("Missing FIXED_IV in .env")

FIXED_IV = base64.b64decode(iv_b64)
if len(FIXED_IV) != 16:
    raise ValueError("AES IV must be 16 bytes")

def encrypt_value(value: str) -> str:
    """Шифрует string в AES-CBC и возвращает base64-encoded string."""
    if value is None:
        raise ValueError("Cannot encrypt None value")
    cipher = AES.new(KEY, AES.MODE_CBC, FIXED_IV)
    padded = pad(str(value).encode(), AES.block_size)
    encrypted = cipher.encrypt(padded)
    return base64.b64encode(encrypted).decode()

def decrypt_value(encrypted_base64: str) -> str:
    """Дешифрует base64-encoded AES-CBC зашифрованную string."""
    encrypted = base64.b64decode(encrypted_base64)
    cipher = AES.new(KEY, AES.MODE_CBC, FIXED_IV)
    decrypted = unpad(cipher.decrypt(encrypted), AES.block_size)
    return decrypted.decode()

File: server/crypto_utils.js
const crypto = require('crypto');
require('dotenv').config();

const key = Buffer.from(process.env.AES_SECRET_KEY, 'base64'); // 32 байта
const fixedIv = Buffer.from(process.env.FIXED_IV, 'base64');   // 16 байт

if (key.length !== 32) throw new Error('AES_SECRET_KEY must be 32 bytes (base64-encoded)');
if (fixedIv.length !== 16) throw new Error('FIXED_IV must be 16 bytes (base64-encoded)');

function encryptValue(value) {
    if (value == null) throw new Error('Cannot encrypt null or undefined');
    const cipher = crypto.createCipheriv('aes-256-cbc', key, fixedIv);
    let encrypted = cipher.update(pad(value.toString()));
    encrypted = Buffer.concat([encrypted, cipher.final()]);
    return encrypted.toString('base64');
}

function decryptValue(encryptedBase64) {
    const encrypted = Buffer.from(encryptedBase64, 'base64');
    const decipher = crypto.createDecipheriv('aes-256-cbc', key, fixedIv);
    let decrypted = decipher.update(encrypted);
    decrypted = Buffer.concat([decrypted, decipher.final()]);
    return unpad(decrypted).toString();
}

// PKCS7 padding to match Python's `Crypto.Util.Padding.pad/unpad`
function pad(bufferStr) {
    const buffer = Buffer.from(bufferStr, 'utf8');
    const blockSize = 16;
    const padding = blockSize - (buffer.length % blockSize);
    const paddingBuffer = Buffer.alloc(padding, padding);
    return Buffer.concat([buffer, paddingBuffer]);
}

function unpad(buffer) {
    const padding = buffer[buffer.length - 1];
    return buffer.slice(0, buffer.length - padding);
}

module.exports = {
    encryptValue,
    decryptValue
};
